src/Java.Runtime.Environment/Java.Interop/ManagedValueManager.cs

```diff
@@ -0,0 +1,237 @@
+using System;
+using System.Collections.Generic;
+using System.Collections.ObjectModel;
+using System.Diagnostics;
+using System.IO;
+using System.Linq;
+using System.Reflection;
+using System.Runtime.CompilerServices;
+using System.Runtime.InteropServices;
+using System.Threading;
+
+namespace Java.Interop {
+
+	class ManagedValueManager : JniRuntime.JniValueManager {
+
+		Dictionary<int, List<IJavaPeerable>>?   RegisteredInstances = new Dictionary<int, List<IJavaPeerable>>();
+
+		public override void WaitForGCBridgeProcessing ()
+		{
+		}
+
+		public override void CollectPeers ()
+		{
+			if (RegisteredInstances == null)
+				throw new ObjectDisposedException (nameof (MonoRuntimeValueManager));
+
+			var peers = new List<IJavaPeerable> ();
+
+			lock (RegisteredInstances) {
+				foreach (var ps in RegisteredInstances.Values) {
+					foreach (var p in ps) {
+						peers.Add (p);
+					}
+				}
+				RegisteredInstances.Clear ();
+			}
+			List<Exception>? exceptions = null;
+			foreach (var peer in peers) {
+				try {
+					peer.Dispose ();
+				}
+				catch (Exception e) {
+					exceptions = exceptions ?? new List<Exception> ();
+					exceptions.Add (e);
+				}
+			}
+			if (exceptions != null)
+				throw new AggregateException ("Exceptions while collecting peers.", exceptions);
+		}
+
+		public override void AddPeer (IJavaPeerable value)
+		{
+			if (RegisteredInstances == null)
+				throw new ObjectDisposedException (nameof (MonoRuntimeValueManager));
+
+			var r = value.PeerReference;
+			if (!r.IsValid)
+				throw new ObjectDisposedException (value.GetType ().FullName);
+			var o = PeekPeer (value.PeerReference);
+			if (o != null)
+				return;
+
+			if (r.Type != JniObjectReferenceType.Global) {
+				value.SetPeerReference (r.NewGlobalRef ());
+				JniObjectReference.Dispose (ref r, JniObjectReferenceOptions.CopyAndDispose);
+			}
+			int key = value.JniIdentityHashCode;
+			lock (RegisteredInstances) {
+				List<IJavaPeerable> peers;
+				if (!RegisteredInstances.TryGetValue (key, out peers)) {
+					peers = new List<IJavaPeerable> () {
+						value,
+					};
+					RegisteredInstances.Add (key, peers);
+					return;
+				}
+
```

Comment:
Do we still need to keep it locked after this point?

Reply:
We need to be locked on *something* at this point: `peers` is still potentially itself referencable by my multiple threads, e.g. if another thread comes in "at the same time" with the same `value.JniIdentityHashCode` value, you could have two threads trying to hit the same `List<IJavaPeerable>`.

We could plausibly instead lock on the `List<IJavaPeerable>`, but this doesn't instill me with confidence either: what happens if one thread is calling `.AddPeer()` while another thread calls `.RemovePeer()` for the same instance?  If we tried "lock `RegisteredInstances`, find the List, unlock `RegisteredInstances`, lock the List, mutate the list", I'm not fully confident that this wouldn't break.

I find it easier to comprehend this way. :-)

Reply:
Then we can start locking the list just before releasing the dictionary (with using `Monitor` directly). Maybe it is not so critical here though.